### PR TITLE
Support `#[macro_reexport]`ing custom derives

### PR DESCRIPTION
**Pull request by jseyfried**
This is gated behind `#![feature(macro_reexport)]` and `#![feature(proc_macro)]`.
r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
cc #37542 #35900
cc @alexcrichton

**Comment by alexcrichton** [comment]
Nice!

File: src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs [hunk 1]
```diff
@@ -1,16 +0,0 @@
...
-// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
-// option. This file may not be copied, modified, or distributed
-// except according to those terms.
-
-// aux-build:derive-a.rs
-
-extern crate derive_a;
-//~^ ERROR: crates of the `proc-macro` crate type cannot be linked at runtime
-
-fn main() {}
```
Comment (nrc) [inline src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs]: This test should still be valid, no?

File: src/test/compile-fail-fulldeps/proc-macro/shadow.rs [hunk 2]
```diff
@@ -15,6 +15,6 @@
 #[macro_use]
 extern crate derive_a;
 #[macro_use]
-extern crate derive_a; //~ ERROR `derive_a` has already been defined
+extern crate derive_a; //~ ERROR `derive_a` has already been imported
 
 fn main() {}
```
Comment (nrc) [inline src/test/compile-fail-fulldeps/proc-macro/shadow.rs]: Should we have a test that uses macro_reexport with a custom derive?

File: src/librustdoc/visit_ast.rs [hunk 3]
```diff
@@ -198,7 +199,11 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
                     if def_id.krate == LOCAL_CRATE {
                         continue // These are `krate.exported_macros`, handled in `self.visit()`.
                     }
-                    let def = self.cx.sess().cstore.load_macro(def_id, self.cx.sess());
+                    let def = match self.cx.sess().cstore.load_macro(def_id, self.cx.sess()) {
+                        LoadedMacro::MacroRules(macro_rules) => macro_rules,
+                        LoadedMacro::ProcMacro(..) => continue, // FIXME
```
Comment (nrc) [inline src/librustdoc/visit_ast.rs]: This FIXME needs more explanation

**Review by nrc — CHANGES_REQUESTED** [review]
I'm not clear on what is going on here - why do we want macro_reexport and custom derive to work together? I would like macro_reexport to disappear entirely and I don't know of any motivation for it to work with custom derive. I assume this is a step towards something else, but I'm not sure what

File: src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs (hunk 1 above)
Comment (jseyfried) [inline src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs]: Is there a good reason to disallow it? For example, we will want to support:
```rust
extern crate derive_a;
use derive_a::A;
#[derive(A)] struct Foo;
```

File: src/librustdoc/visit_ast.rs (hunk 3 above)
Comment (jseyfried) [inline src/librustdoc/visit_ast.rs]: Will do (the FIXME is that `rustdoc` should document custom derives)

**Comment by jseyfried** [comment]
@nrc The main purpose of this PR is to change the metadata to support reexporting custom derives.
Once we support this, I see no reason to specifically *disallow* reexporting custom derives (it would actually be a couple of extra lines).

`#[macro_reexport]` is feature gated and I think we should deprecate it once we can `pub use` macros behind a feature gate (writing that PR now). Until `#[macro_reexport]` is removed, though, think it should behave as consistently as possible on different kinds of macros.

File: src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs (hunk 1 above)
Comment (nrc) [inline src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs]: I would like to maintain a strict phase distinction between macro and non-macro crates (we intend to relax this eventually, but I don't want it to creep in). I don't want the macro_use to be optional, and until we actually support `use`ing derives, it seems that we should require it.

**Comment by nrc** [comment]
> Until #[macro_reexport] is removed, though, think it should behave as consistently as possible on different kinds of macros.

OK, that makes sense.

File: src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs (hunk 1 above)
Comment (jseyfried) [inline src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs]: There is still a phase distinction -- only the macros are available from `extern crate derive_a` and `derive_a` is not linked.

> until we actually support `use`ing derives

I was going to support this behind a feature gate in a PR tomorrow...

> I don't want the macro_use to be optional

Why not? `#[macro_use]` or the ability to `use` derives, `extern crate derive_a;` just defines an empty module `derive_a`. Maybe a warning would be appropriate, but I don't see a reason to disallow it.

Comment (jseyfried) [inline src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs]: There is still a phase distinction -- only the macros are available from `extern crate derive_a` and `derive_a` is not linked.

> until we actually support `use`ing derives

I was going to support this behind a feature gate in a PR tomorrow...

> I don't want the macro_use to be optional

Why not? Without `#[macro_use]` or the ability to `use` derives, `extern crate derive_a;` just defines an empty module `derive_a`. Maybe a warning would be appropriate, but I don't see a reason to disallow it.

Comment (nrc) [inline src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs]: OK, if `use`ing derives will work soon, then lets allow it. As long as the non-macro_use case has sensible semantics, then I see no reason to block it.

Comment (jseyfried) [inline src/test/compile-fail-fulldeps/proc-macro/cannot-link.rs]: Actually, it's already an "unused extern crate" warning.

File: src/librustdoc/visit_ast.rs (hunk 3 above)
Comment (jseyfried) [inline src/librustdoc/visit_ast.rs]: Done.

File: src/test/compile-fail-fulldeps/proc-macro/shadow.rs (hunk 2 above)
Comment (jseyfried) [inline src/test/compile-fail-fulldeps/proc-macro/shadow.rs]: Yeah, I added a test.

**Comment by nrc** [comment]
r = me, but a test seems to be broken

**Review by nrc — APPROVED** [review]

**Comment by jseyfried** [comment]
Travis is broken.

**Pull request merged**